### DISCUSSION TITLE
[RFC] Lua API

### DISCUSSION THREAD

**Discussion by xiyaowong** [vscode-neovim/vscode-neovim#1515] [Ideas]
## Events

Add event module to replace https://github.com/vscode-neovim/vscode-neovim/blob/effad4aaf84a45b770a7a5dceb2a5afdbbf8beca/src/neovim_events_processable.ts, similar to `events.on('redraw')` and `events.on('window_changed')`, combined with type hints, would make it more convenient and easier to maintain.

## Lua API

1. ~Keep the global variable name `vscode`~ Use global variable `__vscode`

	For compatibility reasons, the module name will still be `vscode-neovim`.

2. `vscode.action`

```lua
---Run action, this function is asynchronous
---@param name string The action name, generally a vscode command
---@param opts? table Optional options table, all fields are optional
---            - args: (table) Optional arguments for the action
---            - range: (table) Specific range for the action. In visual mode, this parameter is generally not required.
---            - callback: (function(err: string|nil, ret: any))
---                        Optional callback function to handle the action result.
---                        The first argument is the error message, and the second is the result.
---                        If no callback is provided, any error message will be shown as a notification in VSCode.
---            - enter_insert: (boolean) Whether to run the action in insert mode.
---                            Useful for commands like 'quickfix' or 'code actions',
---                            which may enter the snippet mode in VSCode.
function vscode.action(name, opts) end
```

3. `vscode.action_sync`

```lua
--- Run action, this function is synchronous and blocking
---@param name string The action name, generally a vscode command
---@param opts? table Optional options table, all fields are optional
---            - args: (table) Optional arguments for the action
---            - range: (table) Specific range for the action. In visual mode, this parameter is generally not required.
---            - enter_insert: (boolean) Whether to run the action in insert mode.
---                            Useful for commands like 'quickfix' or 'code actions',
---                            which may enter the snippet mode in VSCode.
---            - timeout: (integer) Timeout in milliseconds. The default value is -1, which means no timeout.
---
---@return string|nil: The error message
---@return any: action result
function vscode.action_sync(name, opts, timeout) end
```

4. `vscode.get_config`

```lua
---Get vscode configuration
---@param name string The full configuration name, such as `editor.tabSize`
---@return any
function vscode.get_config(name) end
```

5. `vscode.update_config`

```lua
---Update vscode configuration
---@param name string The full configuration name, such as `editor.tabSize`
---@param value any
---@param global? boolean If true, updates global settings; otherwise, updates workspace settings.
function vscode.update_config(name, value, global) end
```

6. `vscode.on`

```lua
---Event hooks
---@param event "change_window_state" The event name
---@param callback function The callback function that takes the event arguments as parameters
function vscode.on(event, callback) end
```

7. `vscode.notify`

Like `vim.notify`, but use VSCode notifications

8. `vscode.ui.*` (Low priority)

Like `vim.ui.*`

**Comment by theol0403** [comment]
1.
I don't think there should be a global `vscode` variable, there isn't one now, people can just do `local vscode = require("vscode-neovim")`.

2.
I think we should keep the terminology `call`/`notify`, as it mirrors the RPC/neovim internals and vimscript API. 

I think maybe we should use "command" instead of "name".

How is `range` specified? We need to be able to specify charwise or linewise range. It might even be neat to be able to specify `{start_char = ".", end = "v"}` or something, where `start/end_char: number|string` is charwise and `start/end: number|string` is linewise. 

What is the purpose of `enter_insert`? Can users just call `:startinsert` first? Is it because it includes logic to wait until after `InsertEnter`? 

6.
What is your plan to create vscode events? The VSCode API doesn't really provide events other than callbacks like `onDocumentChange` for a specific buffer. 

7. 
It would be super cool if we could request theme colors to use in highlight groups.

**Comment by xiyaowong** [comment]
It is more convenient for global variables to be called by extension source code or user.

**Reply by theol0403** [reply]
Can you provide an example? I just don't want to pollute any namespace, I know there are people who already have `vscode` as a global in their configs.

**Reply by xiyaowong** [reply]
First, there is no impact on users who use local variables. And I don't think many users have customized the global variable vscode in their own configuration.

**Reply by justinmk** [reply]
Creating a global is a very discouraged practice in Lua and Nvim plugins. Only `vim` is global and that came from from the old vim if_lua interface.

> First, there is no impact on users who use local variables.

It's an unnecessary source of "stuff you need to know". If other plugins do it, the problem compounds.

**Reply by xiyaowong** [reply]
Okay, using `__vscode` is also possible, which can reduce a lot of `require('`

**Reply by xiyaowong** [reply]
Another reason is that I don't want to maintain too many `require` statements https://github.com/LuaLS/lua-language-server/wiki/Libraries unless the user is restricted to only `require('vscode-neovim')`.... Furthermore, even though the context is nvim, vscode-neovim can also be considered as a runtime environment. As far as I know, there are currently only a few `vscode` theme plugins related to nvim plugins, and users won't even need these plugins when using vscode-neovim. Anyway, I can accept using `__vscode`.

**Reply by xiyaowong** [reply]
![image](https://github.com/vscode-neovim/vscode-neovim/assets/47070852/069a6aad-0af6-4dae-acf5-2a168781ce6f)

I would even suggest changing the module name to vscode, just kidding.

**Reply by theol0403** [reply]
`__vscode` is fine, but I don't understand the advantage over simply `require('vscode-neovim')`

See https://github.com/vscode-neovim/vscode-neovim/issues/1354

**Reply by xiyaowong** [reply]
> but I don't understand the advantage over simply `require('vscode-neovim')`

just kidding

I kown this issue. At that time, I had not participated in the contribution of the project, otherwise I would directly tell him to modify his configuration.😗

**Reply by justinmk** [reply]
> which can reduce a lot of require('

`require()` is idiomatic, and fully optimized. So again, there is zero reason to introduce a magic global (including `__vscode`).

**Comment by xiyaowong** [comment]
> I think we should keep the terminology call/notify, as it mirrors the RPC/neovim internals and vimscript API.

There is no need, the user can directly call the vim function.

> I think maybe we should use "command" instead of "name".

Action is not just used to call vscode command.

> How is range specified? We need to be able to specify charwise or linewise range. It might even be neat to be able to specify {start_char = ".", end = "v"} or something, where start/end_char: number|string is charwise and start/end: number|string is linewise.

Lsp range 

> What is the purpose of enter_insert? Can users just call :startinsert first? Is it because it includes logic to wait until after InsertEnter?

Just for convenience. Directly using startinsert will generally cause problems, and we can help ordinary users deal with some details. This parameter doesn't matter to me

**Reply by justinmk** [reply]
> > I think we should keep the terminology call/notify, as it mirrors the RPC/neovim internals and vimscript API.
> 
> There is no need, the user can directly call the vim function.

Familiar conventions can help user intuition. In the case of a Lua api the conventions should probably follow Nvim rather than vscode/javascript, since the context is Nvim.

**Reply by xiyaowong** [reply]
My suggestion is to deprecate the previous Lua interface functions such as notify and notify_range because they are not user-friendly. Can keep it, but prompt the user that it is deprecated.

**Comment by theol0403** [comment]
I'm happy with this RFC, we can start working towards it. The changes make sense to me.

**Comment by xiyaowong** [comment]
I plan to replace the current two methods, vscode-command and vscode-neovim, with a single method, such as vscode-action. From the current perspective, vscode-neovim and vscode-command are redundant, and the current implementation cannot effectively call vscode methods to retrieve data in nvim. For vscode-action, we can either implement ActionManager ourselves or directly use vscode.commands.

https://github.com/vscode-neovim/vscode-neovim/blob/3ccecf6881aca143c195370be884338a59932e77/src/main_controller.ts#L243
https://github.com/vscode-neovim/vscode-neovim/blob/3ccecf6881aca143c195370be884338a59932e77/src/main_controller.ts#L259

Such as https://github.com/vscode-neovim/vscode-neovim/blob/3ccecf6881aca143c195370be884338a59932e77/src/buffer_manager.ts#L213, it should be a action

Such as https://github.com/vscode-neovim/vscode-neovim/blob/3ccecf6881aca143c195370be884338a59932e77/src/buffer_manager.ts#L300, it should be a event

**Reply by justinmk** [reply]
> From the current perspective, vscode-neovim and vscode-command are redundant

Yes, agreed.
